hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/CryptoTransferExecutor.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.token.impl.handlers.transfer;
+
+import static com.hedera.hapi.node.base.ResponseCodeEnum.NOT_SUPPORTED;
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.token.CryptoTransferTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.token.impl.validators.CryptoTransferValidator;
+import com.hedera.node.app.service.token.records.CryptoTransferRecordBuilder;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.HandleException;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import com.hedera.node.config.data.LazyCreationConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+
+public class CryptoTransferExecutor {
+
+    /**
+     * Executes all crypto transfer steps.
+     *
+     * @param txn transaction body
+     * @param transferContext transfer context
+     * @param context handle context
+     * @param validator crypto transfer validator
+     * @param recordBuilder record builder
+     */
+    public static void executeCryptoTransfer(
+            TransactionBody txn,
+            TransferContextImpl transferContext,
+            HandleContext context,
+            CryptoTransferValidator validator,
+            CryptoTransferRecordBuilder recordBuilder) {
+        executeCryptoTransfer(txn, transferContext, context, validator, recordBuilder, false);
+    }
+
+    public static void chargeCustomFee(TransactionBody txn, TransferContextImpl transferContext) {
+        final var customFeeStep = new CustomFeeAssessmentStep(txn.cryptoTransferOrThrow());
+        var transferBodies = customFeeStep.assessCustomFees(transferContext);
+        var topLevelPayer = transferContext.getHandleContext().payer();
+        for (var customFeeTransactionBody : transferBodies) {
+            if (customFeeTransactionBody
```

Comment:
Won't this always be equivalent to skipping the first body,
```
for (int i = 1, n = transferBodies.size(); i < n; i++) {
   ...
```
? I think that will be clearer.

Reply:
Yes, it is the first one.
I was thinking that, if we just skip the first one was not clear enough, why we skip it.
Anyway, I did it like you suggest, and added a comment about it.